### ISSUE TITLE
Reproduce result on CUB and Product datasets

### ISSUE THREAD

**Issue by lugiavn** [lugiavn/generalization-dml#1]
Somebody emailed me about this

**Comment by lugiavn** [comment]
I think I used different learning rate scheduling for CUB because it's easier to overfit:

# learning rate scheduling
if it < 100:
optimizer.param_groups[0]['lr'] = 0.00
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 3000:
optimizer.param_groups[0]['lr'] = 0.01
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 10000:
optimizer.param_groups[0]['lr'] = 0.001
optimizer.param_groups[1]['lr'] = 0.001
optimizer.param_groups[2]['lr'] = 0.001
else:
optimizer.param_groups[0]['lr'] = 0.0001
optimizer.param_groups[1]['lr'] = 0.0001
optimizer.param_groups[2]['lr'] = 0.0001

**Comment by lugiavn** [comment]
I think I used a different learning rate scheduling for CUB because sit's easier to overfit:


if it < 100:
optimizer.param_groups[0]['lr'] = 0.00
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 3000:
optimizer.param_groups[0]['lr'] = 0.01
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 10000:
optimizer.param_groups[0]['lr'] = 0.001
optimizer.param_groups[1]['lr'] = 0.001
optimizer.param_groups[2]['lr'] = 0.001
else:
optimizer.param_groups[0]['lr'] = 0.0001
optimizer.param_groups[1]['lr'] = 0.0001
optimizer.param_groups[2]['lr'] = 0.0001

**Comment by lugiavn** [comment]
For product, the setting is:


batch_size = 64
test_frequency = 5000
loss_frequency = 100

if it < 1000:
optimizer.param_groups[0]['lr'] = 0.00
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 10000:
optimizer.param_groups[0]['lr'] = 0.01
optimizer.param_groups[1]['lr'] = 0.01
optimizer.param_groups[2]['lr'] = 0.01
elif it <= 50000:
optimizer.param_groups[0]['lr'] = 0.003
optimizer.param_groups[1]['lr'] = 0.003
optimizer.param_groups[2]['lr'] = 0.003
elif it <= 100000:
optimizer.param_groups[0]['lr'] = 0.001
optimizer.param_groups[1]['lr'] = 0.001
optimizer.param_groups[2]['lr'] = 0.001
else:
optimizer.param_groups[0]['lr'] = 0.0001
optimizer.param_groups[1]['lr'] = 0.0001
optimizer.param_groups[2]['lr'] = 0.0001

**Comment by lugiavn** [comment]
Code for reading the CUB dataset:

dataset_path = '.'
imid2path = {}
imid2label = {}

with open(dataset_path + '/images.txt') as f:
  lines = f.readlines()
for line in lines:
  line = line.split()
  imid2path[line[0]] = line[1]
with open(dataset_path + '/image_class_labels.txt') as f:
  lines = f.readlines()
for line in lines:
  line = line.split()
  imid2labels[line[0]] = line[1]

train_imgs = []
train_labels = []
test_imgs = []
test_labels = []
for k in imid2path.keys():
  l = int(imid2label[k])
  p = data_path + '/images/' + imid2path[k]
  if l <= 100:
    train_imgs.append(p)
    train_labels.append(l)
  else:
    test_imgs.append(p)
    test_labels.append(l)

**Comment by lugiavn** [comment]
Code for reading CUB dataset
```

dataset_path = '.'
imid2path = {}
imid2label = {}

with open(dataset_path + '/images.txt') as f:
  lines = f.readlines()

for line in lines:
  line = line.split()
  imid2path[line[0]] = line[1]

with open(dataset_path + '/image_class_labels.txt') as f:
  lines = f.readlines()

for line in lines:
  line = line.split()
  imid2labels[line[0]] = line[1]

train_imgs = []
train_labels = []
test_imgs = []
test_labels = []

for k in imid2path.keys():
  l = int(imid2label[k])
  p = data_path + '/images/' + imid2path[k]
  if l <= 100:
    train_imgs.append(p)
    train_labels.append(l)
  else:
    test_imgs.append(p)
    test_labels.append(l)


print 'Done reading data'
print np.min(train_labels)
print np.max(test_labels)
print str(len(train_imgs)) + ' images for training'
print str(len(test_imgs)) + ' images for testing'
```

**Comment by lugiavn** [comment]
Code for reading product dataset

`dataset_path = '/nethome/nvo9/dml17/prod'

train_imgs = []
test_imgs = []
train_labels = []
test_labels = []

with open(dataset_path + '/Stanford_Online_Products/Ebay_train.txt') as f:
    lines = f.readlines()
lines.remove(lines[0])
for line in lines:
    line = line.split()
    train_imgs.append(dataset_path + '/Stanford_Online_Products/' + line[3])
    train_labels.append(int(line[1]))

with open(dataset_path + '/Stanford_Online_Products/Ebay_test.txt') as f:
    lines = f.readlines()
lines.remove(lines[0])
for line in lines:
    line = line.split()
    test_imgs.append(dataset_path + '/Stanford_Online_Products/' + line[3])
    test_labels.append(int(line[1]))
`

**Comment by lugiavn** [comment]
Code for reading product dataset

```
dataset_path = '/nethome/nvo9/dml17/prod'

train_imgs = []
test_imgs = []
train_labels = []
test_labels = []

with open(dataset_path + '/Stanford_Online_Products/Ebay_train.txt') as f:
    lines = f.readlines()
lines.remove(lines[0])
for line in lines:
    line = line.split()
    train_imgs.append(dataset_path + '/Stanford_Online_Products/' + line[3])
    train_labels.append(int(line[1]))

with open(dataset_path + '/Stanford_Online_Products/Ebay_test.txt') as f:
    lines = f.readlines()
lines.remove(lines[0])
for line in lines:
    line = line.split()
    test_imgs.append(dataset_path + '/Stanford_Online_Products/' + line[3])
    test_labels.append(int(line[1]))


print 'Done reading data'
print np.min(train_labels)
print np.max(test_labels)
print str(len(train_imgs)) + ' images for training'
print str(len(test_imgs)) + ' images for testing'

```
